Refactor and stabilize Send/Receive tests

myieye opened the pull request.

hahn-kev commented on backend/Testing/Fixtures/SendReceiveFixture.cs:
```diff
@@ -0,0 +1,62 @@
+        zip.ExtractToDirectory(_templateRepo.FullName);
+    }
+
+    public ProjectConfig InitLocalFlexProjectWithRepo([CallerMemberName] string projectName = "")
```
I moved this here as previously it was a static method in the Utils class, however it only works if this fixture is used, otherwise the template repo folder will be empty. Moving the function here means that it can't be used incorrectly.

hahn-kev reviewed (commented):
left some feedback

rmunn commented:
For what it's worth, when I run this I get four failures, which are really two tests that fail in the same way with both of their input parameters:

SendNewProject fails twice:

```
  Failed Testing.SyncReverseProxy.SendReceiveServiceTests.SendNewProject(totalSizeMb: 180, fileCount: 10) [1 ms]
  Error Message:
   System.IO.DirectoryNotFoundException : Could not find a part of the path '/tmp/SR_Tests/SendNewProject/14-1709b92e/send-new-project-81432895-dev-flex'.
  Stack Trace:
     at Interop.ThrowExceptionForIoErrno(ErrorInfo errorInfo, String path, Boolean isDirError)
   at Interop.CheckIo(Int64 result, String path, Boolean isDirError)
   at System.IO.FileStatus.SetUnixFileMode(SafeFileHandle handle, String path, UnixFileMode mode)
   at LexCore.Utils.FileUtils.CopyFilesRecursively(DirectoryInfo source, DirectoryInfo target, Nullable`1 permissions) in /home/rmunn/code/lexbox/backend/LexCore/Utils/FileUtils.cs:line 18
   at Testing.Fixtures.SendReceiveFixture.InitLocalFlexProjectWithRepo(ProjectPath projectPath) in /home/rmunn/code/lexbox/backend/Testing/Fixtures/SendReceiveFixture.cs:line 51
   at Testing.Fixtures.SendReceiveFixture.InitLocalFlexProjectWithRepo(String projectName) in /home/rmunn/code/lexbox/backend/Testing/Fixtures/SendReceiveFixture.cs:line 45
   at Testing.SyncReverseProxy.SendReceiveServiceTests.SendNewProject(Int32 totalSizeMb, Int32 fileCount) in /home/rmunn/code/lexbox/backend/Testing/SyncReverseProxy/SendReceiveServiceTests.cs:line 245
```

```
  Failed Testing.SyncReverseProxy.SendReceiveServiceTests.SendNewProject(totalSizeMb: 10, fileCount: 1) [< 1 ms]
  Error Message:
   System.IO.DirectoryNotFoundException : Could not find a part of the path '/tmp/SR_Tests/SendNewProject/15-5404b9e7/send-new-project-45b8f40b-dev-flex'.
  Stack Trace:
     at Interop.ThrowExceptionForIoErrno(ErrorInfo errorInfo, String path, Boolean isDirError)
   at Interop.CheckIo(Int64 result, String path, Boolean isDirError)
   at System.IO.FileStatus.SetUnixFileMode(SafeFileHandle handle, String path, UnixFileMode mode)
   at LexCore.Utils.FileUtils.CopyFilesRecursively(DirectoryInfo source, DirectoryInfo target, Nullable`1 permissions) in /home/rmunn/code/lexbox/backend/LexCore/Utils/FileUtils.cs:line 18
   at Testing.Fixtures.SendReceiveFixture.InitLocalFlexProjectWithRepo(ProjectPath projectPath) in /home/rmunn/code/lexbox/backend/Testing/Fixtures/SendReceiveFixture.cs:line 51
   at Testing.Fixtures.SendReceiveFixture.InitLocalFlexProjectWithRepo(String projectName) in /home/rmunn/code/lexbox/backend/Testing/Fixtures/SendReceiveFixture.cs:line 45
   at Testing.SyncReverseProxy.SendReceiveServiceTests.SendNewProject(Int32 totalSizeMb, Int32 fileCount) in /home/rmunn/code/lexbox/backend/Testing/SyncReverseProxy/SendReceiveServiceTests.cs:line 245
```

And ModifyProjectData fails twice:

```
  Failed Testing.SyncReverseProxy.SendReceiveServiceTests.ModifyProjectData(protocol: Hgweb) [1 ms]
  Error Message:
   System.IO.DirectoryNotFoundException : Could not find a part of the path '/tmp/SR_Tests/ModifyProjectData/17-a683ca36/modify-project-data-fb8b4286-dev-flex'.
  Stack Trace:
     at Interop.ThrowExceptionForIoErrno(ErrorInfo errorInfo, String path, Boolean isDirError)
   at Interop.CheckIo(Int64 result, String path, Boolean isDirError)
   at System.IO.FileStatus.SetUnixFileMode(SafeFileHandle handle, String path, UnixFileMode mode)
   at LexCore.Utils.FileUtils.CopyFilesRecursively(DirectoryInfo source, DirectoryInfo target, Nullable`1 permissions) in /home/rmunn/code/lexbox/backend/LexCore/Utils/FileUtils.cs:line 18
   at Testing.Fixtures.SendReceiveFixture.InitLocalFlexProjectWithRepo(ProjectPath projectPath) in /home/rmunn/code/lexbox/backend/Testing/Fixtures/SendReceiveFixture.cs:line 51
   at Testing.Fixtures.SendReceiveFixture.InitLocalFlexProjectWithRepo(String projectName) in /home/rmunn/code/lexbox/backend/Testing/Fixtures/SendReceiveFixture.cs:line 45
   at Testing.SyncReverseProxy.SendReceiveServiceTests.ModifyProjectData(HgProtocol protocol) in /home/rmunn/code/lexbox/backend/Testing/SyncReverseProxy/SendReceiveServiceTests.cs:line 80
```

```
  Failed Testing.SyncReverseProxy.SendReceiveServiceTests.ModifyProjectData(protocol: Resumable) [< 1 ms]
  Error Message:
   System.IO.DirectoryNotFoundException : Could not find a part of the path '/tmp/SR_Tests/ModifyProjectData/18-1d3103c9/modify-project-data-4463aa0b-dev-flex'.
  Stack Trace:
     at Interop.ThrowExceptionForIoErrno(ErrorInfo errorInfo, String path, Boolean isDirError)
   at Interop.CheckIo(Int64 result, String path, Boolean isDirError)
   at System.IO.FileStatus.SetUnixFileMode(SafeFileHandle handle, String path, UnixFileMode mode)
   at LexCore.Utils.FileUtils.CopyFilesRecursively(DirectoryInfo source, DirectoryInfo target, Nullable`1 permissions) in /home/rmunn/code/lexbox/backend/LexCore/Utils/FileUtils.cs:line 18
   at Testing.Fixtures.SendReceiveFixture.InitLocalFlexProjectWithRepo(ProjectPath projectPath) in /home/rmunn/code/lexbox/backend/Testing/Fixtures/SendReceiveFixture.cs:line 51
   at Testing.Fixtures.SendReceiveFixture.InitLocalFlexProjectWithRepo(String projectName) in /home/rmunn/code/lexbox/backend/Testing/Fixtures/SendReceiveFixture.cs:line 45
   at Testing.SyncReverseProxy.SendReceiveServiceTests.ModifyProjectData(HgProtocol protocol) in /home/rmunn/code/lexbox/backend/Testing/SyncReverseProxy/SendReceiveServiceTests.cs:line 80
```

myieye commented on backend/Testing/Services/UtilsTests.cs:
```diff
@@ -0,0 +1,17 @@
+{
+    [Theory]
+    [InlineData(nameof(SendReceiveServiceTests.ModifyProjectData), "modify-project-data")]
+    [InlineData(nameof(SendReceiveServiceTests.SendNewProject), "send-new-project")]
```
Because I want to be able to generate project codes so that:
1) They're somewhat standardized
2) I don't have to think about them ever again (that's the best case scenario 🤣)

It's a bit goofy having a test to test a test util. I know. But it's so tiny.

myieye commented:
I merged it so that we can get some feedback.

- We now have a couple repo/project utils that make it easy to initialize seperate repos per test.
- I did a pretty big refactor on `SendReceiveAfterProjectReset` partially so that I can run it in Windows 🤷 and so that it uses the new repo utils I introuced, so it's a lot shorter now.
- I added a second set of parameters for `SendNewProject` so that we also send a smaller one. That way we have higher confidence if the big one fails, but the smaller one passes
- I've changed the `ModifyProjectData` test so that it uses its won repo instead of sena-3. The hope is that that will reduce some race conditions 🙏.
- I standardized how we explicitly wait for hg's `refreshinterval` in `Utils.WaitForHgRefreshIntervalAsync()`
